buzzmobile/tests/test_utils/mock_nodes.py

```diff
@@ -102,16 +101,20 @@ def cb_wrapper(message):
         test_node.received = True
         test_node._message = message  # pylint: disable=protected-access
 
-        return new_callback(message)
-
-    rospy.Subscriber(topic, rosmsg_type, cb_wrapper)
-    yield test_node
-    rospy.signal_shutdown('test complete')
-
-    # Ros really doesn't want you to reinitialize a node once it's been
-    # shutdown because there can be bad side effects, but we are good
-    # at cleaning up after ourselves.
-    rospy.client._init_node_args = None  # pylint: disable=protected-access
-    rospy.core._shutdown_flag = False  # pylint: disable=protected-access
-    rospy.core._in_shutdown = False  # pylint: disable=protected-access
+        return callback(message)
+
+    sub = rospy.Subscriber(topic, rosmsg_type, cb_wrapper)
+    try:
+        yield test_node
+    finally:
+        sub.unregister()
+        rospy.signal_shutdown('test complete')
+        reload(rospy)
+
+        # Ros really doesn't want you to reinitialize a node once it's been
+        # shutdown because there can be bad side effects, but we are good
+        # at cleaning up after ourselves.
+        rospy.client._init_node_args = None  # pylint: disable=protected-access
+        rospy.core._shutdown_flag = False  # pylint: disable=protected-access
+        rospy.core._in_shutdown = False  # pylint: disable=protected-access
```

Comment:
Interestingly, with the xdist based testing, none of this cleanup may be needed.